### PR TITLE
Wip/hughsie/pixart

### PR DESCRIPTION
**Pull request by hughsie**
Type of pull request:
- [ ] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [X] Code fix
- [ ] Feature
- [ ] Documentation

### CODE REVIEW

File: libfwupdplugin/fu-udev-device.c [hunk 1]
```diff
@@ -1108,6 +1108,93 @@ fu_udev_device_set_physical_id (FuUdevDevice *self, const gchar *subsystems, GEr
...
 
+/**
+ * fu_udev_device_set_logical_id:
+ * @self: A #FuUdevDevice
+ * @subsystems: A subsystem string, e.g. `pci,usb`
+ * @error: A #GError, or %NULL
+ *
+ * Sets the logical ID from the device subsystem. Plugins should choose the
+ * subsystem that is "deepest" in the udev tree, for instance choosing 'usb'
+ * over 'pci' for a mouse device.
```
Comment (superm1) [inline libfwupdplugin/fu-udev-device.c:1119]: your examples here aren't used, maybe make it choosing `hid` over `usb`

**Review by superm1 — APPROVED** [review]

**Comment by hughsie** [comment]
Lets see if PixArt gets back to us about why the HID report is sometimes prefixed. I don't like papering over bugs.

**Comment by superm1** [comment]
>Lets see if PixArt gets back to us about why the HID report is sometimes prefixed. I don't like papering over bugs.

Yeah that does look weird.

**Comment by jyu99** [comment]
Hi 

I found the some information about HIDGETFEATURE
[https://www.kernel.org/doc/Documentation/hid/hidraw.txt](url)

and here

[https://stackoverflow.com/questions/62333326/zero-byte-disappear-when-get-hid-feature-report-using-ioctl-in-linux](url)

 It's seem like skip the first byte, I guess default  first byte is report number.

But in your case. not only skip the first byte, then system will keep report number automatically

**Comment by hughsie** [comment]
> This commit: bluez/bluez@35a2c50 present in BlueZ

Good detective work!

> I guess we should include some code to detect the BlueZ version.

I agree. It needs to be a runtime daemon-version test rather than a built-time test as the user is free to update fwupd and bluez independently. I suspect the best place might be in fu-bluez-backend.c and then set a "backend-version" string property and fu_bluez_device_get_backend_version() in FuBluezDevice. Actually, it might make sense in FuDevice (the parent class) too -- we could do the same for the udev version and gusb version -- the latter actually being useful in bugreports. Could you do that in a new PR please?

**Comment by hughsie** [comment]
> FuPxiDevice is not a FuBluezDevice

Urgh, I forgot that bit. Can we do both? i.e. do the bluez-backend thing (which I think we need anyway for this kind of thing) and then we just copy and paste the code into FuPxiDevice. Or we do the bluez-backend thing and then do something like my hack i suggested -- but instead of shuffling the bytes along for new versions, prepending the bytes for old versions....

**Comment by hughsie** [comment]
I think we'll need to support the older bluez as it's in the LTS releases of various things (including RHEL IIRC)

**Comment by superm1** [comment]
>I think we'll need to support the older bluez as it's in the LTS releases of various things (including RHEL IIRC)

But also there is the possibility to have a runtime check on the bluez version and disable plugins if it's too old.  We do this in other plugins (modem manager comes to mind).

File: libfwupdplugin/fu-udev-device.c (hunk 1 above)
Comment (hughsie) [reply libfwupdplugin/fu-udev-device.c:1119]: Fixed. I've also done `s\subsystems\subsystem` as I don't think the fallback makes sense for the logical ID. You know what you're looking for, unlike the physical ID.

**Pull request merged**